File: fabric-audit-agent/core/detectors/pipeline.js
import { DEFAULT_CONFIG } from '../config.js';

/** @typedef {import('./capacity.js').Flag} Flag */

/**
 * Refresh/pipeline health detectors. Pure: facts in, flags out.
 * @param {{pipelines?:object[]}} facts
 * @param {object} [config]
 * @returns {Flag[]}
 */
export function detectPipelines(facts, config = DEFAULT_CONFIG) {
  const pipelines = facts?.pipelines ?? [];
  const flags = [];
  for (const p of pipelines) {
    const where = `${p.workspace} / ${p.name}`;
    if (p.lastStatus === 'Failed' || (p.failRatePct ?? 0) >= config.pipeline.failRatePct) {
      flags.push({
        type: 'pipeline.failing', resource: where, when: p.lastRunAt ?? '',
        evidence: { status: p.lastStatus, failRatePct: p.failRatePct ?? 0 },
        what: `Pipeline "${p.name}" last status ${p.lastStatus} (fail rate ${p.failRatePct ?? 0}%).`,
      });
    }
    if (p.gatewayHealthy === false) {
      flags.push({
        type: 'pipeline.gateway', resource: where, when: p.lastRunAt ?? '',
        evidence: {},
        what: `Pipeline "${p.name}" depends on an unhealthy gateway.`,
      });
    }
  }
  return flags;
}

File: fabric-audit-agent-app/fabric_audit_agent/detectors/pipeline.py
"""Pipeline/refresh-health detectors. Faithful port of the Node ``core/detectors/pipeline.js``."""
from ..config import DEFAULT_CONFIG


def detect_pipelines(facts, config=None):
    config = config or DEFAULT_CONFIG
    pipelines = (facts or {}).get("pipelines") or []
    thr = config["pipeline"]
    flags = []
    for p in pipelines:
        where = f"{p.get('workspace')} / {p.get('name')}"
        when = p.get("lastRunAt") or ""
        if p.get("lastStatus") == "Failed" or (p.get("failRatePct") or 0) >= thr["failRatePct"]:
            flags.append({
                "type": "pipeline.failing", "resource": where, "when": when,
                "evidence": {"status": p.get("lastStatus"), "failRatePct": p.get("failRatePct") or 0},
                "what": f"Pipeline \"{p.get('name')}\" last status {p.get('lastStatus')} (fail rate {p.get('failRatePct') or 0}%).",
            })
        if p.get("gatewayHealthy") is False:
            flags.append({
                "type": "pipeline.gateway", "resource": where, "when": when, "evidence": {},
                "what": f"Pipeline \"{p.get('name')}\" depends on an unhealthy gateway.",
            })
    return flags
